File: client/jetbrains/src/main/java/com/sourcegraph/telemetry/PostStartupActivity.java
package com.sourcegraph.telemetry;

import com.intellij.ide.plugins.IdeaPluginDescriptor;
import com.intellij.ide.plugins.PluginInstaller;
import com.intellij.ide.plugins.PluginStateListener;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupActivity;
import com.sourcegraph.cody.CodyAgentProjectListener;
import com.sourcegraph.config.ConfigUtil;
import com.sourcegraph.config.SettingsChangeListener;
import java.util.UUID;
import org.jetbrains.annotations.NotNull;

public class PostStartupActivity implements StartupActivity.DumbAware {
  private static String generateAnonymousUserId() {
    return UUID.randomUUID().toString();
  }

  @Override
  public void runActivity(@NotNull Project project) {
    // Make sure that SettingsChangeListener is loaded
    project.getService(SettingsChangeListener.class);

    // When no anonymous user ID is set yet, we create a new one and treat this as an installation
    // event.
    // This likely means that the user has never started IntelliJ with our extension before
    if (ConfigUtil.getAnonymousUserId() == null) {
      ConfigUtil.setAnonymousUserId(generateAnonymousUserId());
    }

    PluginInstaller.addStateListener(
        new PluginStateListener() {
          public void install(@NotNull IdeaPluginDescriptor ideaPluginDescriptor) {
            CodyAgentProjectListener.startAgent(project);
            GraphQlLogger.logInstallEvent(project)
                .thenAccept(
                    wasSuccessful -> {
                      if (wasSuccessful) {
                        ConfigUtil.setInstallEventLogged(true);
                      }
                    });
          }

          @Override
          public void uninstall(@NotNull IdeaPluginDescriptor ideaPluginDescriptor) {
            CodyAgentProjectListener.stopAgent(project);
            if (ideaPluginDescriptor
                .getPluginId()
                .getIdString()
                .equals("com.sourcegraph.jetbrains")) {
              GraphQlLogger.logUninstallEvent(project);

              // Clearing this so that we can detect a new installation if the user re-enables the
              // extension.
              ConfigUtil.setAnonymousUserId(null);

              ConfigUtil.setInstallEventLogged(false);
            }
          }
        });
  }
}

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/initialization/PostStartupActivity.kt
package com.sourcegraph.cody.initialization

import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.StartupActivity
import com.sourcegraph.cody.agent.CodyAgentManager
import com.sourcegraph.cody.auth.SelectOneOfTheAccountsAsActive
import com.sourcegraph.cody.config.SettingsMigration
import com.sourcegraph.cody.statusbar.CodyAutocompleteStatusService
import com.sourcegraph.config.CodyAuthNotificationActivity
import com.sourcegraph.config.ConfigUtil
import com.sourcegraph.telemetry.TelemetryInitializerActivity

/**
 * StartupActivity is obsolete in recent platform versions.
 *
 * TODO: We should migrate to com.intellij.openapi.startup.ProjectActivity when we bump
 *   compatibility.
 */
class PostStartupActivity : StartupActivity.DumbAware {
  override fun runActivity(project: Project) {
    TelemetryInitializerActivity().runActivity(project)
    SettingsMigration().runActivity(project)
    SelectOneOfTheAccountsAsActive().runActivity(project)
    CodyAuthNotificationActivity().runActivity(project)
    if (ConfigUtil.isCodyEnabled()) CodyAgentManager.startAgent(project)
    CodyAutocompleteStatusService.resetApplication(project)
  }
}
